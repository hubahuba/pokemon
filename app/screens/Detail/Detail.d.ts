import {PokemonData} from '@/core/definitions/data/pokemon.definition';
import {StackNavigation} from '~/routes/index.d';
import {BerryData, PokemonStat} from '@/definitions/usecases/pokemon';

export interface DetailViewModelProps {
  data: PokemonData;
  navigation: StackNavigation;
  iChooseU: (data: PokemonData) => void;
  deletePokemon: () => void;
  berries?: BerryData[] | string;
  feedPokemon: (ownedId: string, berry: BerryData) => void;
  evolutionPokemon: ({pokemonId}: {pokemonId: string}) => void;
  setShowBerry: (show: boolean) => void;
  showBerry: boolean;
}

export interface StatViewProps {
  name: string;
  weight?: number | null;
  maxWeight?: number | null;
  stats: PokemonStat[];
  owned?: boolean;
  ownedId?: string;
  onDeletePokemon?: () => void;
  onEvolution?: (ownedId: string) => void;
}

export interface PokemonBerriesProps {
  ownedId: string;
  data: BerryData[];
  onCloseBerry?: () => void;
  onSelectBerry?: (ownedId: string, berri: BerryData) => void;
}
