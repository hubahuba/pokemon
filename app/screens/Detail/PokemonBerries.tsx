import * as React from 'react';
import {View} from 'react-native';
import Animated, {
  FadeInUp,
  FadeOutDown,
  interpolate,
} from 'react-native-reanimated';
import Carousel from 'react-native-reanimated-carousel';
import {GestureHandlerRootView} from 'react-native-gesture-handler';

import IconButton from '~atoms/IconButton';
import {PokemonBerriesProps} from '~/screens/Detail/Detail';
import {memo, useState} from 'react';
import BerryButton from '~atoms/BerryButton';

const PokemonBerries = memo(function PokemonBerries({
  data,
  onCloseBerry,
  onSelectBerry,
  ownedId,
}: PokemonBerriesProps) {
  const [viewWidth, setViewWidth] = useState(0);
  const itemSize = 80;
  const iconSize = 34;
  const centerOffset = viewWidth / 2 - itemSize / 2;

  const animationStyle = React.useCallback(
    (value: number) => {
      'worklet';

      const itemGap = interpolate(
        value,
        [-3, -2, -1, 0, 1, 2, 3],
        [-30, -10, 0, 0, 0, 10, 30],
      );

      const translateX =
        interpolate(value, [-1, 0, 1], [-itemSize, 0, itemSize]) +
        centerOffset -
        itemGap;

      const translateY = interpolate(
        value,
        [-1, -0.3, 0, 0.4, 1],
        [50, 35, 30, 35, 50],
      );

      const scale = interpolate(
        value,
        [-1, -0.5, 0, 0.5, 1],
        [0.8, 0.85, 1.1, 0.85, 0.8],
      );

      return {
        transform: [
          {
            translateX,
          },
          {
            translateY,
          },
          {scale},
        ],
      };
    },
    [centerOffset],
  );

  if (!ownedId) {
    return <View />;
  }

  return (
    <Animated.View
      testID="PokemonBerriesWrapper"
      onLayout={event => setViewWidth(event.nativeEvent.layout.width)}
      entering={FadeInUp}
      exiting={FadeOutDown}>
      <GestureHandlerRootView>
        <Carousel
          width={itemSize}
          height={itemSize}
          style={{
            width: '100%',
            height: viewWidth / 2,
          }}
          loop
          autoPlay={false}
          data={data}
          renderItem={({item}) => (
            <BerryButton
              item={item}
              onPress={onSelectBerry}
              ownedId={ownedId}
              testID="SelectBerry"
            />
          )}
          customAnimation={animationStyle}
        />
        <IconButton
          testID="CloseBerriesButton"
          className="absolute rounded-full bg-gray-400 p-1 border border-gray-500"
          style={{
            top: viewWidth / 2 - iconSize,
            left: viewWidth / 2 - iconSize / 2,
          }}
          iconName="close"
          onPress={onCloseBerry}
        />
      </GestureHandlerRootView>
    </Animated.View>
  );
});

export default PokemonBerries;
