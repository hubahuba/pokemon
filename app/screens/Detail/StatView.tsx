import * as React from 'react';
import {View, Text, TouchableOpacity} from 'react-native';
import PokemonStatText from '~/components/molecules/PokemonStatDisplay/index.tsx';
import {StatViewProps} from '~/screens/Detail/Detail';
import IconButton from '~atoms/IconButton';

export default function StatView({
  name,
  weight,
  maxWeight,
  stats,
  owned,
  ownedId,
  onDeletePokemon,
  onEvolution,
}: StatViewProps) {
  return (
    <View className="bg-gray-100 p-4 rounded w-full">
      <View className="flex border-b-gray-300 border-b pb-1 mb-2 items-center">
        {owned && (
          <IconButton
            color="#FF0000"
            className="absolute top-2 right-0"
            iconName="delete"
            onPress={() => onDeletePokemon?.()}
          />
        )}
        <Text className="font-[Raleway-Bold] text-[28px]">{name}</Text>
      </View>
      <View className="w-full">
        {stats.map(stat => {
          return (
            <View key={stat.name} className="mb-2">
              <PokemonStatText
                name={stat.name as string}
                value={stat.value / 255}
              />
            </View>
          );
        })}
        {(weight as number) >= (maxWeight as number) && ownedId ? (
          <View>
            <TouchableOpacity
              onPress={() => {
                if (ownedId) {
                  onEvolution?.(ownedId);
                }
              }}
              className="bg-amber-500 px-4 py-2 rounded justify-center items-center">
              <Text className="font-[Raleway-Bold] text-[18px] text-white">
                Evolution
              </Text>
            </TouchableOpacity>
          </View>
        ) : (
          <View>
            <PokemonStatText
              name="Weight"
              value={(weight as number) / (maxWeight as number)}
            />
          </View>
        )}
      </View>
    </View>
  );
}
