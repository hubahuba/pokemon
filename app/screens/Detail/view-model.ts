import {RouteProp, useNavigation, useRoute} from '@react-navigation/native';
import {Popup} from 'react-native-popup-confirm-toast';
import {StackNavigation, RootStackParamList} from '~/routes/routes';
import {DetailViewModelProps} from './Detail';
import services from '@/services.ts';
import {BerryData} from '@/definitions/usecases/pokemon';
import {useState} from 'react';

export default function ViewModel(): DetailViewModelProps {
  const route = useRoute<RouteProp<RootStackParamList, 'Detail'>>();
  const navigation = useNavigation<StackNavigation>();
  const [showBerry, setShowBerry] = useState(false);
  const {data} = route.params;

  let berries: BerryData[] | string | undefined =
    services.storage.getItem('berries');
  if (berries && typeof berries === 'string') {
    berries = JSON.parse(berries);
  }

  const iChooseU = () => {
    const newData = services.useCase.gameAction.iChooseYou(data);
    navigation.navigate('Detail', {data: newData});
  };

  const deletePokemon = () => {
    Popup.show({
      // @ts-ignore
      type: 'confirm',
      title: 'Hapus Pokemon!',
      textBody: 'Yakin hapus pokemon ini?',
      buttonText: 'Ya, hapus',
      confirmText: 'Batal',
      icon: require('~/assets/images/pikachu-sad.png'),
      okButtonTextStyle: {
        fontFamily: 'Raleway-Bold',
      },
      okButtonStyle: {
        backgroundColor: '#ff0000',
      },
      confirmButtonTextStyle: {
        fontFamily: 'Raleway-Bold',
      },
      titleTextStyle: {
        fontFamily: 'Raleway-Bold',
      },
      descTextStyle: {
        fontFamily: 'Raleway-Medium',
      },
      callback: () => {
        Popup.hide();
        setTimeout(() => {
          services.useCase.gameAction.deletePokemon(data.ownedId);
          navigation.goBack();
        }, 300);
      },
      cancelCallback: () => {
        Popup.hide();
      },
    });
  };

  const feedPokemon = (ownedId: string, berry: BerryData) => {
    const pokemon = services.useCase.gameAction.feedPokemon(ownedId, berry);
    if (pokemon) {
      navigation.navigate('Detail', {data: pokemon});
    }
  };

  const evolutionPokemon = ({pokemonId}: {pokemonId: string}) => {
    setShowBerry(false);
    const current = services.useCase.gameAction.evolutionPokemon(pokemonId);
    if (current) {
      navigation.navigate('Detail', {data: current});
    }
  };

  return {
    data,
    navigation,
    iChooseU,
    deletePokemon,
    berries,
    showBerry,
    setShowBerry,
    feedPokemon,
    evolutionPokemon: evolutionPokemon,
  };
}
