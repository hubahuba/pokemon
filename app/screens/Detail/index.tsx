import * as React from 'react';
import {ImageBackground, View} from 'react-native';
import FastImage from 'react-native-fast-image';
import Animated, {FadeInUp, FadeOutDown} from 'react-native-reanimated';

import StatView from '~/screens/Detail/StatView';
import IconButton from '~/components/atoms/IconButton';
import ViewModel from '~/screens/Detail/view-model';
import ImageButton from '~atoms/ImageButton';
import PokemonBerries from '~/screens/Detail/PokemonBerries';
import {BerryData, PokemonData} from '@/definitions/usecases/pokemon';

function Detail() {
  const {
    data,
    navigation,
    iChooseU,
    deletePokemon,
    berries,
    showBerry,
    setShowBerry,
    feedPokemon,
    evolutionPokemon,
  } = ViewModel();
  if (data) {
    const {
      name,
      image,
      baseWeight,
      currentWeight,
      stats,
      owned,
      maxWeight,
      nextEvolution,
      ownedId,
    } = data as PokemonData;

    return (
      <ImageBackground
        className="items-center min-h-screen"
        style={{width: '100%', height: '100%'}}
        source={require('~/assets/images/background.jpg')}
        blurRadius={10}>
        <View className="flex-1 flex flex-col basis-11/12">
          <IconButton
            className="rounded-full bg-gray-400 p-1 absolute right-0 top-3 border border-gray-500"
            iconName="close"
            onPress={() => navigation.goBack()}
          />

          <View className="grow">
            <FastImage
              defaultSource={require('~/assets/images/pokeball.png')}
              source={{uri: image}}
              style={{width: 340, height: 340}}
            />
            {stats && (
              <StatView
                name={name}
                weight={owned === true ? currentWeight : baseWeight}
                maxWeight={maxWeight}
                stats={stats}
                owned={owned}
                ownedId={ownedId}
                onDeletePokemon={deletePokemon}
                onEvolution={currentOwnedId =>
                  evolutionPokemon({pokemonId: currentOwnedId})
                }
              />
            )}
          </View>

          {showBerry && berries !== undefined && (
            <PokemonBerries
              onCloseBerry={() => setShowBerry(false)}
              data={berries as BerryData[]}
              ownedId={ownedId as string}
              onSelectBerry={feedPokemon}
            />
          )}

          {owned !== true && (
            <View className="flex-none items-center">
              <ImageButton
                source={require('~/assets/images/choose.png')}
                onPress={() => iChooseU(data)}
              />
            </View>
          )}

          {owned && (
            <Animated.View
              entering={FadeInUp}
              exiting={FadeOutDown}
              className="flex-none items-center">
              <ImageButton
                source={require('~/assets/images/berry.png')}
                onPress={() => setShowBerry(true)}
              />
            </Animated.View>
          )}
        </View>
      </ImageBackground>
    );
  }
}

export default Detail;
