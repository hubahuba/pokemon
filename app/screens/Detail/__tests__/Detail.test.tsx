import React from 'react';
import {Root as PopupRootProvider} from 'react-native-popup-confirm-toast';
import {waitFor, screen, render} from '@testing-library/react-native';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import Detail from '~/screens/Detail';

describe('Detail view-model', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        gcTime: Infinity,
      },
    },
  });

  const wrapper = ({children}: {children: any}) => (
    <QueryClientProvider client={queryClient}>
      {/*
      // @ts-ignore */}
      <PopupRootProvider>{children}</PopupRootProvider>
    </QueryClientProvider>
  );

  test('Detail screen ', async () => {
    render(<Detail />, {wrapper});
    await waitFor(() => {
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
