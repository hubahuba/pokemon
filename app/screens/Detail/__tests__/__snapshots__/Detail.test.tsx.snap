// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail view-model Detail screen  1`] = `
[
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      [
        {
          "alignItems": "center",
        },
        {
          "minHeight": 1334,
        },
        {
          "height": "100%",
          "width": "100%",
        },
      ]
    }
  >
    <Image
      blurRadius={10}
      source={
        {
          "testUri": "../../../app/assets/images/background.jpg",
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        [
          {
            "flexBasis": "0%",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          {
            "display": "flex",
          },
          {
            "flexDirection": "column",
          },
          {
            "flexBasis": "91.666667%",
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#9ca3af",
            "borderBottomColor": "#6b7280",
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderBottomWidth": 1,
            "borderLeftColor": "#6b7280",
            "borderLeftWidth": 1,
            "borderRightColor": "#6b7280",
            "borderRightWidth": 1,
            "borderTopColor": "#6b7280",
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
            "borderTopWidth": 1,
            "opacity": 1,
            "paddingBottom": 4,
            "paddingLeft": 4,
            "paddingRight": 4,
            "paddingTop": 4,
            "position": "absolute",
            "right": 0,
            "top": 12,
          }
        }
        testID="DetailBackButton"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": undefined,
                "fontSize": 24,
              },
              undefined,
              {
                "fontFamily": "anticon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          î˜³
        </Text>
      </View>
      <View
        style={
          [
            {
              "flexGrow": 1,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "height": 340,
                "width": 340,
              },
            ]
          }
        >
          <FastImageView
            defaultSource={
              {
                "testUri": "../../../app/assets/images/pokeball.png",
              }
            }
            resizeMode="cover"
            style={
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
        0
      </View>
      <View
        style={
          [
            {
              "flexBasis": "auto",
              "flexGrow": 0,
              "flexShrink": 0,
            },
            {
              "alignItems": "center",
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "height": 120,
                  "width": 120,
                },
              ]
            }
          >
            <FastImageView
              defaultSource={null}
              resizeMode="contain"
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "transform": [
          {
            "translateY": 1334,
          },
        ],
        "width": 750,
        "zIndex": 10,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "minHeight": 0,
          "overflow": "hidden",
          "position": "absolute",
          "transform": [
            {
              "translateY": 1334,
            },
          ],
          "width": "90%",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "#fff",
              "height": 75,
              "width": 100,
            },
            false,
          ]
        }
      />
      <Image
        resizeMode="contain"
        source={
          {
            "testUri": "../../react-native-popup-confirm-toast/src/assets/warning.png",
          }
        }
        style={
          {
            "height": 48,
            "position": "absolute",
            "top": 20,
            "width": 48,
          }
        }
      />
      <View
        style={
          {
            "padding": 20,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#111111",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "400",
                "lineHeight": 24,
                "textAlign": "center",
              },
              false,
            ]
          }
        />
        <View
          style={false}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#702c91",
                "borderRadius": 8,
                "flex": 1,
                "height": 48,
                "justifyContent": "center",
                "marginTop": 20,
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                    "fontSize": 16,
                    "fontStyle": "normal",
                    "fontWeight": "600",
                    "lineHeight": 20,
                    "textAlign": "center",
                  },
                  false,
                ]
              }
            >
              Ok
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": undefined,
        "borderRadius": 0,
        "borderWidth": 0,
        "display": "none",
        "elevation": 5,
        "flexDirection": "row",
        "minHeight": 120,
        "position": "absolute",
        "shadowColor": "#1da1f2",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
        "transform": [
          {
            "translateY": -1334,
          },
        ],
        "width": 750,
        "zIndex": 99999,
      }
    }
  >
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 0,
        }
      }
    >
      <View
        style={
          [
            {
              "flex": 1,
              "marginTop": 0,
              "paddingLeft": 20,
              "paddingRight": 20,
            },
          ]
        }
      />
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#122459",
            "height": 5,
            "position": "absolute",
            "top": -10,
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      />
    </View>
  </View>,
]
`;
